File: apps/admin-dashboard/lib/api/api-error-envelope.ts
export type ApiErrorEnvelope = {
  readonly code: string;
  readonly message: string;
  readonly details?: unknown;
};

export function parseApiErrorEnvelope(value: unknown): ApiErrorEnvelope {
  if (!isObject(value)) {
    return {
      code: 'UNKNOWN_ERROR',
      message: 'Unknown API error',
    };
  }

  return {
    code: typeof value.code === 'string' ? value.code : 'UNKNOWN_ERROR',
    message:
      typeof value.message === 'string'
        ? value.message
        : 'Unknown API error',
    details: value.details,
  };
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: apps/mobile/lib/core/networking/api_error_envelope.dart
class ApiErrorEnvelope {
  const ApiErrorEnvelope({
    required this.code,
    required this.message,
    this.details,
  });

  factory ApiErrorEnvelope.fromJson(Map<String, dynamic> json) {
    return ApiErrorEnvelope(
      code: json['code'] as String? ?? 'UNKNOWN_ERROR',
      message: json['message'] as String? ?? 'Unknown API error',
      details: json['details'],
    );
  }

  final String code;
  final String message;
  final Object? details;
}
